### PR TITLE
[dart][dart-dio] Use built_value collection types without string replacement

### PR DESCRIPTION
**Pull request by kuhnroyal**
Do not use String replacement for built_value collection types. 
This fixes https://github.com/OpenAPITools/openapi-generator/issues/4865 but also changes return/parameter types and would be a breaking change.
I think this is the correct thing to do but the changed parameter types may reduce usability in cases of list body/parameters.

@josh-burton Did you intentionally prevent the return/parameter types from being built_value types? What is your opinion on this?

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [x] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

CC @ircecho (2017/07) @swipesight (2018/09) @jaumard (2018/09) @josh-burton (2019/12) @amondnet (2019/12) @sbu-WBT (2020/12) @kuhnroyal (2020/12) @agilob (2020/12)

### CODE REVIEW

**Review by josh-burton — APPROVED** [review]

**Comment by josh-burton** [comment]
@kuhnroyal no not intentional, more a case of not knowing how to correctly substitute the built collection types. Even though a breaking change I think it's the right change. Nice work :)

**Comment by kuhnroyal** [comment]
@wing328 Is master the correct branch for a breaking change? How should this be documented?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java [hunk 1]
```diff
@@ -188,6 +177,8 @@ public DartClientCodegen() {
         importMapping.put("Set", "dart:core");
         importMapping.put("DateTime", "dart:core");
 
+        defaultIncludes = new HashSet<>(Collections.singletonList("dart:core"));
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java:180]: `Sets.newHashSet` from guava

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java [hunk 2]
```diff
@@ -474,34 +465,62 @@ public String toDefaultValue(Schema schema) {
 
     @Override
     public String getTypeDeclaration(Schema p) {
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java:467]: can you add unit tests to this method?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java [hunk 3]
```diff
@@ -474,34 +465,62 @@ public String toDefaultValue(Schema schema) {
...
     }
 
     @Override
     public Map<String, Object> postProcessModels(Map<String, Object> objs) {
         return postProcessModelsEnum(objs);
     }
 
+    @Override
+    public CodegenOperation fromOperation(String path, String httpMethod, Operation operation, List<Server> servers) {
+        final CodegenOperation op = super.fromOperation(path, httpMethod, operation, servers);
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java:509]: can you add unit tests for this method? is it worth to make it more dart-dio specific as the rest of the changes are dart-dio?

**Comment by agilob** [comment]
I'm not a fan of these changes TBH, it has the following disadvantages I see straight away:

1. Breaking change in terms of input/output to methods
2. 3rd party library to support such basic concepts like lists and maps
3. I heave to know how to use that 3rd party library to use lists/maps (I was not aware this library even exists until now)
4. When I generate an Owner with Pets I want to add and remove pets from existing array as I go and not be limited by  immutable arrays that it holds by default. It restricts how classes can be used.
5. TypeScript and Java generators that I use produce mutable lists so it's also introducing incompatibility between languages and how generated code can be used

If anything, this should be a runtime flag in custom properties.

That said, I much prefer shelf and not planing to use dart-dio, so whatever floats your boat works for me, but please don't put it in main dart. If these changes landed in any generator I use I would be really pissed as it completely breaks 100% of cases how I use generated classes in Java and Dart. Autogenerated code are my DTOs and small cached objects, so shared schema and how classes are used between languages absolutely has to be the same, eg. lists and maps have to be mutable.

#4865 should be solved in a way that's compatible with other generators, so contributors from scala or python can come to the dart generators and be familiars with code generation process.

**Comment by agilob** [comment]
Any opinion on this @sbu-WBT ?

**Comment by kuhnroyal** [comment]
> When I generate an Owner with Pets I want to add and remove pets from existing array as I go and not be limited by immutable arrays that it holds by default. It restricts how classes can be used.

IMO this depends on the language/framework features. I usually prefer immutability. Many other generators for languages that have immutable collections by default do the same. Java and Typescript are not some of those as we all know :)

If you choose to use this generator, you also choose to use `built_value`, all the models you get from the API are immutable built_value classes, everything you put into it as well. It only makes sense to use the immutable data structures provided by built_value which use the same builder pattern as the model classes. It is much easier for responses which are deserialized as `BuiltList/Map/Set` anyways.

> but please don't put it in main dart

I would never dream of it. This is here for dart-dio because the used collection classes are part of the same serialization package that is used.

> It takes dart-dio generated code further from dart generator, making them less and less compatible. TypeScript is suffering from similar problems, each generator is completely different, has different codes, settings and produces different outputs. Java and Python generators are all similar, you can swap gson with jackson generator and almost no difference in use. We should be aiming for less split between generators, not more.

Sounds great and would very much be possible if we could use dart mirrors in Flutter, like we use reflection in Java :) Languages that don't have a form reflection will have the problem of generators being a lot more different to each other because everything that is different has to be expressed in code and can not be weaved in at runtime.


It is a great idea to provide a flag to choose a specific serialization framework, as other generators do. But dart-dio doesn't do this yet and it is way out of scope for this issue.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java (hunk 2 above)
Comment (kuhnroyal) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java:467]: There are tests for dart and dart-dio that test the type returned for `array` and `map` OAS types.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java (hunk 3 above)
Comment (kuhnroyal) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java:509]: Could possibly be moved to the dart-dio generator.

**Comment by agilob** [comment]
>Sounds great and would very much be possible if we could use dart mirrors in Flutter, like we use reflection in Java :) Languages that don't have a form reflection will have the problem of generators being a lot more different to each other because everything that is different has to be expressed in code and can not be weaved in at runtime.

>You can swap gson with jackson generator and almost no difference in use

I didn't mean reflection or mirrors, just content of X1Generator and X2Generator and .mustache files should be as similar as possible to reduce duplication, produce outputs that are more similar. You can swap gson for jackson and as long as there are correct dependencies, it will all work without any fuss.

But as above, I'm not familiar with dart-dio, so whatever works for you here.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java (hunk 3 above)
Comment (agilob) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java:509]: makes sense to keep it here for all darts.

**Review by agilob — APPROVED** [review]
Minor change requests not worth effort of a commit, good to go.

**Pull request merged**